File: Assignment4/3.py
# Write a program to collect height & weight of N students. Find the highest weight/height
# ratio using linked list.

class Node:
    def __init__(self,data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def insert_end(self,data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            return
        temp = self.head
        while temp.next:
            temp = temp.next
        temp.next = new_node

    def highest_ratio(self):
        temp = self.head
        max_ratio = 0
        max_index = 0
        index = 0
        while temp:
            if temp.data > max_ratio:
                max_ratio = temp.data
                max_index = index
            temp = temp.next
            index += 1
            return max_ratio,max_index+1

n = int(input("Enter number of students: "))
ll = LinkedList()

for i in range(n):
    h,w = map(float,input(f"Enter height and weight of student {i+1}: ").split())
    ll.insert_end(w/h)

ratio,student = ll.highest_ratio()
print(f"Highest weight/height ratio = {ratio:.2f} (Student {student})")

File: Assignment4/3.c
// Write a program to collect height & weight of N students.
// Find the highest weight/height ratio using linked list.
#include <stdio.h>
#include<stdlib.h>

struct Node{
    float data;
    struct Node* next;
};

struct Node* createNode(float data){
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

int main() {
    int n;
    printf("Enter number of students: ");
    scanf("%d", &n);

    struct Node* head = NULL;
    struct Node* tail = NULL;

    for(int i=0;i<n;i++){
        float h,w;
        printf("Enter height and weight of student %d: ",i+1);
        scanf("%f %f",&h,&w);
        float ratio = w/h;
        struct Node* newNode = createNode(ratio);
        if(head == NULL){
            head = newNode;
            tail = newNode;
        }
        else{
            tail->next = newNode;
            tail = newNode;
        }
    }

    struct Node* temp = head;
    float max_ratio = 0;
    int student_index = 0, max_index = 0;
    while(temp != NULL){
        if(temp->data > max_ratio){
            max_ratio = temp->data;
            max_index = student_index;
        }
        temp = temp->next;
        student_index++;
    }
    printf("Heighest weight/height ratio = %.2f (student %d)\n",max_ratio,max_index+1);

    return 0;
}